sys/src/9/port/devramfs.c

```diff
@@ -208,19 +273,44 @@ ramread(Chan *c, void *buf, int32_t n, int64_t off)
 	return n;
 }
 
+typedef double Align;
+typedef union Header Header;
+
+union Header {
+        struct {
+                Header* next;
+                uint    size;
+        } s;
+        Align   al;
+};
+
+
 static int32_t
 ramwrite(Chan* c, void* v, int32_t n, int64_t off)
 {
+//	Header *p;
+
 	qlock(&ramlock);
 	struct RamFile *file = (void*)c->qid.path;
-	if(n+off > file->length){
-		void *newfile = realloc(file->data, n+off);
-		if(newfile == nil){
-			return 0;
+	if (file->magic != RAM_MAGIC)
+		error(INVALID_FILE);
+	if(n+off >= file->length){
+		uint64_t alloclength = file->alloclength;
+		while(alloclength < n + off)
+			alloclength += RAM_BLOCK_LEN;
+		if(alloclength > file->alloclength){
+			void *newfile = realloc(file->data, alloclength);
+			if(newfile == nil){
+				qunlock(&ramlock);
+				return 0;
+			}
+			file->data = newfile;
+			file->alloclength = alloclength;
 		}
-		file->data = newfile;
 		file->length = n+off;
 	}
+//	p = (Header*)file->data - 1;
```

Comment:
suggestion. Rather than comment the code out, control it with it
if (ramfsdebug)

why: the test is cheap, and then while debugging we can turn these off and on in gdb, which is handy. Russ pointed out to me once how much he dislikes compiling out debug code, and I tend to agree. Also, bracketing with an if makes sure this will always compile. If you really hate the variable, maybe bracket it with if (0) and it gets snipped out by good compilers (even ken c would do that) but you're still sure it compiles.